Enable verbose logging when rbd mirroing is enabled for debugging

malayparida2000 opened the pull request:
When mirroring is enabled it creates a enable job which runs the
required ceph config set commands to enable verbose logging.
When mirroring is disabled it creates a disable job which runs the
required ceph config rm commands to disable verbose logging and deletes
both the enable and disable jobs.
RBD team needs by default verbose logging when rbd-mirroing is enabled,
to debug issues properly. This would be reverted once RDR goes to GA.

Resolves-https://bugzilla.redhat.com/show_bug.cgi?id=2127186

malayparida2000 commented:
`sh-4.4$ ceph config dump
WHO                                      MASK  LEVEL     OPTION                                 VALUE                              RO
global                                         basic     log_to_file                            true                                 
global                                         advanced  mon_allow_pool_delete                  true                                 
global                                         advanced  mon_allow_pool_size_one                true                                 
global                                         advanced  mon_cluster_log_file                                                        
global                                         advanced  mon_pg_warn_min_per_osd                0                                    
mon                                            advanced  auth_allow_insecure_global_id_reclaim  false                                
mgr                                            advanced  mgr/balancer/mode                      upmap                                
mgr                                            advanced  mgr/prometheus/rbd_stats_pools         ocs-storagecluster-cephblockpool   * 
mgr                                            advanced  mgr/rbd_support/log_level              debug                                
osd.0                                          basic     osd_mclock_max_capacity_iops_ssd       16840.108369                         
osd.1                                          basic     osd_mclock_max_capacity_iops_ssd       17091.355128                         
osd.2                                          basic     osd_mclock_max_capacity_iops_ssd       17071.832952                         
mds.ocs-storagecluster-cephfilesystem-a        basic     mds_cache_memory_limit                 4294967296                           
mds.ocs-storagecluster-cephfilesystem-a        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
mds.ocs-storagecluster-cephfilesystem-b        basic     mds_cache_memory_limit                 4294967296                           
mds.ocs-storagecluster-cephfilesystem-b        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
client.rbd-mirror-peer                         advanced  debug_ms                               1/1                                  
client.rbd-mirror-peer                         advanced  debug_rbd                              20/20                                
client.rbd-mirror-peer                         advanced  debug_rbd_mirror                       30/30                                
client.rbd-mirror-peer                         basic     log_file                               /var/log/ceph/$cluster-$name.log   * 
client.rbd-mirror.a                            advanced  debug_ms                               1/1                                  
client.rbd-mirror.a                            advanced  debug_rbd                              20/20                                
client.rbd-mirror.a                            advanced  debug_rbd_mirror                       30/30                                
client.rbd-mirror.a                            basic     log_file                               /var/log/ceph/$cluster-$name.log   * 
sh-4.4$ `

idryomov commented on controllers/storagecluster/rbd_mirror_debug.go:
```diff
@@ -0,0 +1,129 @@
+								"-c",
+								`echo "starting configuration of mirror logging"
+								ceph config set client.rbd-mirror.a debug_ms 1
+								ceph config set client.rbd-mirror.a debug_rbd 20
```
```suggestion
								ceph config set client.rbd-mirror.a debug_rbd 15
```

malayparida2000 replied:
ack

idryomov commented on controllers/storagecluster/rbd_mirror_debug.go:
```diff
@@ -0,0 +1,129 @@
+								ceph config set client.rbd-mirror.a debug_rbd_mirror 30
+								ceph config set client.rbd-mirror.a log_file /var/log/ceph/\$cluster-\$name.log
+								ceph config set client.rbd-mirror-peer debug_ms 1
+								ceph config set client.rbd-mirror-peer debug_rbd 20
```
```suggestion
								ceph config set client.rbd-mirror-peer debug_rbd 15
```

idryomov commented on controllers/storagecluster/reconcile.go:
```diff
@@ -172,6 +172,13 @@ func (r *StorageClusterReconciler) Reconcile(ctx context.Context, request reconc
+	// Ensure that verbose logging or debugging is enabled when RBD mirroring is enabled
+	// This is a temporary arrangement, this is to be removed when RDR goes to GA
+	err = r.ensureRbdMirrorDebug(sc)
+	if err != nil && !errors.IsNotFound(err) {
```
Just curious, why is this error ignored here?  What resource is expected to be not found?

malayparida2000 replied:
Thanks for noticing, This is a mistake left from me. I meant to ignore the is already exists error, But seems like I was confused. Will correct this.

Madhu-1 replied:
You should be ignoring the AlreadyExist error, not the notFound here right?

malayparida2000 replied:
Yes, I will set it to ignore the already exists error instead of this not found error

idryomov commented on controllers/storagecluster/rbd_mirror_debug.go:
```diff
@@ -0,0 +1,129 @@
+)
+
+func (r *StorageClusterReconciler) ensureRbdMirrorDebug(sc *ocsv1.StorageCluster) error {
+	if sc.Spec.Mirroring.Enabled {
```
These settings should be mostly harmless to leave on but it would be nice to disable them automatically -- it's just a bunch of `ceph config rm` commands after all.

malayparida2000 replied:
Sounds good, I will create another job that will be run when mirroring is disabled. It will remove the settings we applied with the creation job.

malayparida2000 commented on controllers/storagecluster/rbd_mirror_debug.go:
```diff
@@ -0,0 +1,129 @@
+
+func (r *StorageClusterReconciler) ensureRbdMirrorDebug(sc *ocsv1.StorageCluster) error {
+	if sc.Spec.Mirroring.Enabled {
+		return r.Client.Create(context.TODO(), rbdMirrorDebugJob(sc))
```
Okay, I will create a bool variable that stores if the job creation was already done, So we will call the create func
if (sc.Spec.Mirroring.Enabled  && !rbdMirrorDebugJobCreated), Sounds good?

malayparida2000 replied:
Cool, Will keep that

Madhu-1 commented on controllers/storagecluster/rbd_mirror_debug.go:
```diff
@@ -0,0 +1,129 @@
+	debugJob := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rbd-mirror-debug",
+			Namespace: sc.Namespace,
```
Add ownerRef for this job to get deleted when rook/ocs is uninstalled.

malayparida2000 replied:
Thanks for pointing it out, Will do it.

malayparida2000 commented:
mparida@mparida-mac ~ % oc patch storagecluster $(oc get storagecluster -n openshift-storage -o=jsonpath='{.items[0].metadata.name}')  -n openshift-storage --type json --patch  '[{ "op": "replace", "path": "/spec/mirroring", "value": {"enabled": false} }]'
    storagecluster.ocs.openshift.io/ocs-storagecluster patched
    mparida@mparida-mac ~ % oc exec rook-ceph-tools-79c8b56799-dfrnk -- ceph config dump
    WHO                                      MASK  LEVEL     OPTION                                 VALUE                              RO
    global                                         basic     log_to_file                            true                                 
    global                                         advanced  mon_allow_pool_delete                  true                                 
    global                                         advanced  mon_allow_pool_size_one                true                                 
    global                                         advanced  mon_cluster_log_file                                                        
    global                                         advanced  mon_pg_warn_min_per_osd                0                                    
    mon                                            advanced  auth_allow_insecure_global_id_reclaim  false                                
    mgr                                            advanced  mgr/balancer/mode                      upmap                                
    mgr                                            advanced  mgr/prometheus/rbd_stats_pools         ocs-storagecluster-cephblockpool   * 
    osd.0                                          basic     osd_mclock_max_capacity_iops_ssd       16613.703840                         
    osd.1                                          basic     osd_mclock_max_capacity_iops_ssd       17339.390414                         
    osd.2                                          basic     osd_mclock_max_capacity_iops_ssd       16545.057735                         
    mds.ocs-storagecluster-cephfilesystem-a        basic     mds_cache_memory_limit                 4294967296                           
    mds.ocs-storagecluster-cephfilesystem-a        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
    mds.ocs-storagecluster-cephfilesystem-b        basic     mds_cache_memory_limit                 4294967296                           
    mds.ocs-storagecluster-cephfilesystem-b        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
    mparida@mparida-mac ~ % oc patch storagecluster $(oc get storagecluster -n openshift-storage -o=jsonpath='{.items[0].metadata.name}')  -n openshift-storage --type json --patch  '[{ "op": "replace", "path": "/spec/mirroring", "value": {"enabled": true} }]' 
    storagecluster.ocs.openshift.io/ocs-storagecluster patched
    mparida@mparida-mac ~ % oc exec rook-ceph-tools-79c8b56799-dfrnk -- ceph config dump
    WHO                                      MASK  LEVEL     OPTION                                 VALUE                              RO
    global                                         basic     log_to_file                            true                                 
    global                                         advanced  mon_allow_pool_delete                  true                                 
    global                                         advanced  mon_allow_pool_size_one                true                                 
    global                                         advanced  mon_cluster_log_file                                                        
    global                                         advanced  mon_pg_warn_min_per_osd                0                                    
    mon                                            advanced  auth_allow_insecure_global_id_reclaim  false                                
    mgr                                            advanced  mgr/balancer/mode                      upmap                                
    mgr                                            advanced  mgr/prometheus/rbd_stats_pools         ocs-storagecluster-cephblockpool   * 
    mgr                                            advanced  mgr/rbd_support/log_level              debug                                
    osd.0                                          basic     osd_mclock_max_capacity_iops_ssd       16613.703840                         
    osd.1                                          basic     osd_mclock_max_capacity_iops_ssd       17339.390414                         
    osd.2                                          basic     osd_mclock_max_capacity_iops_ssd       16545.057735                         
    mds.ocs-storagecluster-cephfilesystem-a        basic     mds_cache_memory_limit                 4294967296                           
    mds.ocs-storagecluster-cephfilesystem-a        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
    mds.ocs-storagecluster-cephfilesystem-b        basic     mds_cache_memory_limit                 4294967296                           
    mds.ocs-storagecluster-cephfilesystem-b        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
    client.rbd-mirror-peer                         advanced  debug_ms                               1/1                                  
    client.rbd-mirror-peer                         advanced  debug_rbd                              15/15                                
    client.rbd-mirror-peer                         advanced  debug_rbd_mirror                       30/30                                
    client.rbd-mirror-peer                         basic     log_file                               /var/log/ceph/$cluster-$name.log   * 
    client.rbd-mirror.a                            advanced  debug_ms                               1/1                                  
    client.rbd-mirror.a                            advanced  debug_rbd                              15/15                                
    client.rbd-mirror.a                            advanced  debug_rbd_mirror                       30/30                                
    client.rbd-mirror.a                            basic     log_file                               /var/log/ceph/$cluster-$name.log   * 
    mparida@mparida-mac ~ % oc patch storagecluster $(oc get storagecluster -n openshift-storage -o=jsonpath='{.items[0].metadata.name}')  -n openshift-storage --type json --patch  '[{ "op": "replace", "path": "/spec/mirroring", "value": {"enabled": false} }]'
    storagecluster.ocs.openshift.io/ocs-storagecluster patched
    mparida@mparida-mac ~ % oc exec rook-ceph-tools-79c8b56799-dfrnk -- ceph config dump
    WHO                                      MASK  LEVEL     OPTION                                 VALUE                              RO
    global                                         basic     log_to_file                            true                                 
    global                                         advanced  mon_allow_pool_delete                  true                                 
    global                                         advanced  mon_allow_pool_size_one                true                                 
    global                                         advanced  mon_cluster_log_file                                                        
    global                                         advanced  mon_pg_warn_min_per_osd                0                                    
    mon                                            advanced  auth_allow_insecure_global_id_reclaim  false                                
    mgr                                            advanced  mgr/balancer/mode                      upmap                                
    mgr                                            advanced  mgr/prometheus/rbd_stats_pools         ocs-storagecluster-cephblockpool   * 
    osd.0                                          basic     osd_mclock_max_capacity_iops_ssd       16613.703840                         
    osd.1                                          basic     osd_mclock_max_capacity_iops_ssd       17339.390414                         
    osd.2                                          basic     osd_mclock_max_capacity_iops_ssd       16545.057735                         
    mds.ocs-storagecluster-cephfilesystem-a        basic     mds_cache_memory_limit                 4294967296                           
    mds.ocs-storagecluster-cephfilesystem-a        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
    mds.ocs-storagecluster-cephfilesystem-b        basic     mds_cache_memory_limit                 4294967296                           
    mds.ocs-storagecluster-cephfilesystem-b        basic     mds_join_fs                            ocs-storagecluster-cephfilesystem    
    mparida@mparida-mac ~ %

malayparida2000 commented:
Hi @travisn @idryomov @Madhu-1 , I have addressed all the requested changes.
1. The debug_rbd is now 15 instead of 20
2. The job is created only when mirroring is enabled & the job isn't there
**3. If mirroring is not enabled anymore, The original debug job is deleted, An undo job is run to remove all the ceph configs, Then the undo job is also deleted after completion.**
4. The job has owner reference of the storagecluster.

Please take another look

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+
+func (r *StorageClusterReconciler) ensureRbdMirrorDebug(sc *ocsv1.StorageCluster) error {
+	job := &batchv1.Job{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
```
Instead of context.TODO reuse the `r.ctx` in all places

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+	}, job)
+
+	if sc.Spec.Mirroring.Enabled {
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Create the debug job
+				return r.Client.Create(context.TODO(), rbdMirrorDebugJob(sc))
+			}
+			return err
+		}
+		return nil
```
```suggestion
if err == nil || ! errors.IsNotFound(err){
    return err
 }

// Create the debug job
return r.Client.Create(context.TODO(), rbdMirrorDebugJob(sc))
```

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+		}
+		// Create a undo job to remove the ceph configs applied by the debug job
+		err = r.Client.Create(context.TODO(), undoRbdMirrorDebugJob(sc))
+		if err != nil {
```
you also need to take care of the already existing case here

```suggestion
if err != nil &&  errors.IsAlreadyExist(err){
	return err
}
```

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+			return err
+		}
+		// wait for the undo job to succeed & then delete it
+		err = utilwait.PollImmediate(5*time.Second, 60*time.Second, func() (done bool, err error) {
```
What if this job takes more time complete, like an unstable network case of slow image pulling etc?

malayparida2000 replied:
So you are suggesting I should increase the timeout? to probably 120 sec/ 180 sec/ 300 sec?

Madhu-1 replied:
If it fails for 60 seconds and we are going to retry as Reconcile fails, we are good. If not it should be taken care of.

malayparida2000 replied:
Yes, we will retry as reconcile will fail & would be tried again

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+				Name:      "undo-rbd-mirror-debug",
+				Namespace: sc.Namespace,
+			}, job)
+			if err != nil {
```
Do you need to return true and nil for not found error here?

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+}
+
+func rbdMirrorDebugJob(sc *ocsv1.StorageCluster) *batchv1.Job {
+	rookImage := os.Getenv("ROOK_CEPH_IMAGE")
```
You need to set affinity or toleration to run this on only dedicated nodes, not on any nodes in the cluster.

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +130,283 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+	return debugJob
+}
+
+func undoRbdMirrorDebugJob(sc *ocsv1.StorageCluster) *batchv1.Job {
```
Avoid code duplication by defining a common job for both cases and having different functions to define only ceph arguments.

iamniting commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +134,210 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+			return err
+		}
+		// Wait for the job to succeed
+		err = r.waitForJobCompletion(enableRbdMirrorDebugLoggingJobName, sc.Namespace)
```
I don't think it is a good idea to wait in the reconcile, instead requeue after sometime

malayparida2000 replied:
Yes, Will requeue instead of wait

iamniting commented on controllers/storagecluster/reconcile.go:
```diff
@@ -172,6 +172,13 @@ func (r *StorageClusterReconciler) Reconcile(ctx context.Context, request reconc
 	}
 
+	// Ensure that verbose logging or debugging is enabled when RBD mirroring is enabled
+	// This is a temporary arrangement, this is to be removed when RDR goes to GA
```
Can we add TODO here?

iamniting reviewed (changes requested):
Can you separate this commit into 2 one for generated changes and one for controllers?

malayparida2000 commented:
@iamniting @Madhu-1 , All requested changes have been addressed, PTAL

idryomov reviewed (approved):
Ack from the RBD perspective

iamniting commented on controllers/storagecluster/reconcile.go:
```diff
@@ -172,6 +172,13 @@ func (r *StorageClusterReconciler) Reconcile(ctx context.Context, request reconc
 
+	// Ensure that verbose logging or debugging is enabled when RBD mirroring is enabled
+	// TODO: This is a temporary arrangement, this is to be removed when RDR goes to GA
+	result, err = r.ensureRbdMirrorDebugLogging(sc)
```
I don't think this is the correct place to call this method, We should move this to the reconcilePhases

iamniting commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,194 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+		r.Log.Info("Successfully enabled rbd mirror debug logging")
+		rbdMirrorDebugLoggingEnabled = true
+	}
+	// Mirrooring is not enabled but the debug logging is enabled
```
```suggestion
	// Mirroring is disabled but the debug logging is enabled
```

iamniting commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,194 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+ceph config rm mgr mgr/rbd_support/log_level
+		echo "completed"`
+
+	// Mirrooring is enabled but the debug logging is not enabled
```
```suggestion
	// Mirroring is enabled but the debug logging is disabled
```

malayparida2000 replied:
I am surprised the spell-check test missed this

iamniting commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,194 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+		}
+
+		// Delete both the enable & disable jobs
+		err = r.Client.Delete(r.ctx, getRbdMirrorDebugLoggingJob(sc, disableRbdMirrorDebugLoggingJobName, disableRbdMirrorDebugLoggingCommands))
```
We don't need to call the whole func, We just need a kind with name and namespace set to delete an object.

iamniting commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,194 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+		if err != nil && !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		err = r.Client.Delete(r.ctx, getRbdMirrorDebugLoggingJob(sc, enableRbdMirrorDebugLoggingJobName, enableRbdMirrorDebugLoggingCommands))
```
same here

iamniting commented:
/hold

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,204 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if job.Status.Succeeded == 0 {
+			return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
+		}
```
Add more details/comments about this check. Is this to ensure the job is succeeded? What if the job is not successful at all?

malayparida2000 replied:
done

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,204 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+			return reconcile.Result{}, err
+		}
+		if job.Status.Succeeded == 0 {
+			return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
```
if the job was not successful this will block the storagecluster delete workflow also?

malayparida2000 replied:
IMO No, The function call is at the end, after all the calls regarding storage cluster phase/ deletion are done successfully. So I don't think it will affect any of that.

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,204 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+								{Name: "mon-endpoint-volume", MountPath: "/etc/rook"},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								AllowPrivilegeEscalation: &[]bool{false}[0],
```
use 
```
import "k8s.io/utils/pointer"
pointer.BoolPtr(false)
```
 ?

Madhu-1 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -124,3 +133,204 @@ func (obj *ocsCephRbdMirrors) ensureDeleted(r *StorageClusterReconciler, sc *ocs
+								configCommands,
+							},
+							SecurityContext: &corev1.SecurityContext{
+								AllowPrivilegeEscalation: &[]bool{false}[0],
```
Same here as well use already available package

malayparida2000 commented:
/hold cancel

malayparida2000 commented:
/unhold

malayparida2000 commented:
/cherry-pick release-4.12

malayparida2000 commented on controllers/storagecluster/cephrbdmirrors.go:
```diff
@@ -3,20 +3,29 @@ package storagecluster
 type ocsCephRbdMirrors struct{}
 
+var (
+	rbdMirrorDebugLoggingEnabled = false
```
Even if the var is reset, it will try to create the job again, So there it will encounter an alreadyExists error, Then it will check if the job was successful. If it was, then it will set the var to true again. So I think we are safe.
